processing_control/processing_control.pde

```diff
@@ -1,219 +1,342 @@
-    println(" done.");
-}  
+// This Processing sketch opens a GUI in which users can specify the
+// conductance parameters used by the Teensy microcontroller.
+//
+// There are eight of them at present:
+//     - shunt conductance (g_shunt, nS)
+//     - maximum HCN conductance (g_hcn, nS)
+//     - maximum sodium conductance (g_Na, nS)
+//     - mean excitatory Ornstein-Uhlenbeck conductance (m_OU_exc, nS)
+//     - diffusion constant of excitatory Ornstein-Uhlenbeck conductance (D_OU_exc, nS^2/ms)
+//     - mean inhibitory Ornstein-Uhlenbeck conductance (m_OU_inh, nS)
+//     - diffusion constant of inhibitory Ornstein-Uhlenbeck conductance (D_OU_inh, nS^2/ms)
+//     - maximum EPSC conductance (g_epsc, nS)
+//
+// The numbers can be adjusted using the sliders.
+//
+// Pressing "connect" will connect to the microcontroller via the serial port.
+// Pressing "upload" will send the numbers in the GUI to the microcontroller.
+// Pressing "zero" will set all the numbers to zero and send zeros to the microcontroller.
+//
+// The sketch requires the ControlP5 library.
+//
+
+import controlP5.*;
+import processing.serial.Serial;
+import java.text.*;
+import java.io.*;
+import java.nio.*;
+
+// ---
+// a serial object (the port to communicate with the microcontroller)
+// change port if needed in the UI
+String DEFAULT_SERIAL_PORT = System.getProperty("os.name").startsWith("Windows")
+  ? " COM3"                    // usual windows default
+  : " /dev/cu.usbmodem101";    // usual posix default
+Serial mcuPort;
+
+// ---
+// define variables for the ControlP5 object (the GUI)
+ControlP5 cp5;
+
+color textColor = color(100, 100, 100);
+color inputColor = color(64, 64, 64);
+color disabledColor = color(200, 200, 200);
+color buttonColor = color(0, 45, 95);
+color okColor = color(64, 200, 64);
+color failColor = color(200, 64, 64);
+
+Button connectButton;
+Button uploadButton;
+Button zeroButton;
+Textfield connectTextfield;
+Textfield g_epsc_textfield;
+
+// ---
+// storage vars
+String[] echo = new String[8];
+
+// initialize the variables set by the GUI
+float g_shunt = 0;
+float g_hcn = 0;
+float g_Na = 0;
+float m_OU_exc = 0;
+float D_OU_exc = 0;
+float m_OU_inh = 0;
+float D_OU_inh = 0;
+
+
+void setup() {
+  int offsetX = 100;
+  int offsetY = 50;
+
+  // specify GUI window size, color, and text case
+  size(450, 650);
+  Label.setUpperCaseDefault(false);
+
+  // create the ControlP5 object, add sliders, specify the font, and add buttons
+  cp5 = new ControlP5(this);
+
+  PFont pfont = createFont("Arial", 8, true);
+  ControlFont font = new ControlFont(pfont, 18);
+  cp5.setFont(font);
+
+  connectTextfield = cp5.addTextfield("Serial Port", offsetX, offsetY, 200, 30)
+    .setText(DEFAULT_SERIAL_PORT)
+    .setColorBackground(inputColor);
+  connectButton = cp5.addButton("Connect");
+  connectButton.setPosition(offsetX + 220, offsetY)
+    .setSize(80, 30)
+    .setColorBackground(buttonColor)
+    .setColorForeground(textColor);
```

Comment:
added a field and button to allow the user to specify the serial port at-will and click a button to connect. this prevents the app from hanging on startup if connection fails and provides visual feedback. the field will turn green is successful and red if there was a problem